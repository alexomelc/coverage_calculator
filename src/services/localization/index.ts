const localization = {
  product: {
    yourTariffPrice: 'Your tariff price is:',
    selectCoverage: 'Please, select coverage from range',
    addToBasket: 'Add to basket',
  },
  shoppingCart: {
    shoppingCart: 'Shopping card',
    youHave: ' You have',
    items: 'items',
  },
};

export default localization;