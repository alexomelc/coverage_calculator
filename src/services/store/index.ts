import { configureStore } from '@reduxjs/toolkit';
import shoppingCartSlice from '../redux/shoppingCart';

export const store = configureStore({
  reducer: {
    shoppingCart: shoppingCartSlice,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
