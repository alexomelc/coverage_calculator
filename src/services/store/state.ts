import { ProductItem } from '../types/index.d';

export const INITIAL_STATE: ProductItem[] = [
  {
    id: '1',
    type: 'Bike',
    coverage: {
      min: 0,
      max: 3000,
    },
    risk: 30,
  },
  {
    id: '2',
    type: 'Jewelry',
    coverage: {
      min: 500,
      max: 10000,
    },
    risk: 5,
  },
  {
    id: '3',
    type: 'Electronics',
    coverage: {
      min: 500,
      max: 6000,
    },
    risk: 35,
  },
  {
    id: '4',
    type: 'Sports Equipment',
    coverage: {
      min: 0,
      max: 20000,
    },
    risk: 30,
  },
];
