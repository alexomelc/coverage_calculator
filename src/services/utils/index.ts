export const calculatePriceTariff = (coverage: number, risk: number) =>
  Math.floor((coverage / 100) * risk)
