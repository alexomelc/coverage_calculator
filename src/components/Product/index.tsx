import { useState } from 'react';
import {
  Card,
  CardActions,
  CardContent,
  Button,
  Typography,
} from '@mui/material';
import { addProduct } from '../../services/redux/shoppingCart';
import { useDispatch } from 'react-redux';
import RangePicker from '../RangePicker';
import { calculatePriceTariff } from '../../services/utils';
import localization from '../../services/localization';
import { ProductItem } from '../../services/types/index.d';

interface Props {
  product: ProductItem;
}

const Product = (props: Props) => {
  const [selectedUserCoverage, setSelectedUserCoverage] = useState<
    number | undefined
  >(undefined);
  const dispatch = useDispatch();

  const addProductToShoppingCard = (id: string, calculatedPrice: number) => {
    dispatch(addProduct({ id, calculatedPrice }));
  };

  const {
    product: { type, coverage, risk, id, isProductAdded },
  } = props;

  const calculatedPriceTariff = selectedUserCoverage
    ? calculatePriceTariff(selectedUserCoverage, risk)
    : 0;

  return (
    <Card sx={{ minWidth: '250px' }}>
      <CardContent>
        <Typography
          data-testid="product-type"
          gutterBottom
          variant="h5"
          component="span">
          {type}
        </Typography>
        <Typography variant="body2" color="text.secondary">
          {calculatedPriceTariff
            ? `${localization.product.yourTariffPrice} ${calculatedPriceTariff}`
            : localization.product.selectCoverage}
        </Typography>
        <Typography component="div" variant="body2" color="text.secondary">
          <RangePicker
            setSelectedUserCoverage={setSelectedUserCoverage}
            minCoverage={coverage.min}
            maxCoverage={coverage.max}
            isProductAdded={isProductAdded}
          />
        </Typography>
      </CardContent>
      <CardActions sx={{ justifyContent: 'flex-end' }}>
        <Button
          onClick={() => addProductToShoppingCard(id, calculatedPriceTariff)}
          disabled={isProductAdded || !selectedUserCoverage}
          size="small">
          {localization.product.addToBasket}
        </Button>
      </CardActions>
    </Card>
  );
};

export default Product;
