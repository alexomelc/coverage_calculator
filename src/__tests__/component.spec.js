import {
  render,
  screen,
  fireEvent,
  RenderResult,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import Product from '../components/Product';

import { store } from '../services/store';
import { ProductItem } from '../services/types/index.d';

const renderProduct = (product: ProductItem): RenderResult =>
  render(
    <Provider store={store}>
      <Product product={product} />
    </Provider>
  );


const getProduct = (id: string): ProductItem => {
  const book = store.getState().shoppingCart.find((book) => book.id === id);
  expect(book).not.toBeUndefined();
  return book;
};


test('Renders Product component', () => {
  const productId = '1';
  const product = getProduct(productId);
  const { asFragment } = renderProduct(product);

  const productText = screen.getByTestId('product-type');
  expect(productText).toHaveTextContent('Bike');  

  expect(asFragment()).toMatchSnapshot();
});

