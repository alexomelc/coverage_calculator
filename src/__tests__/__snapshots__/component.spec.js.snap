// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Product component 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1ni1nnj-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <span
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        data-testid="product-type"
      >
        Bike
      </span>
      <p
        class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        Please, select coverage from range
      </p>
      <div
        class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        <div
          class="MuiBox-root css-ax14a8"
        >
          <div
            class="MuiBox-root css-1domaf0"
          >
            <span
              aria-labelledby="input-slider"
              class="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root MuiSlider-marked css-1aznpnh-MuiSlider-root"
            >
              <span
                class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
              />
              <span
                class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
                style="left: 0%; width: 0%;"
              />
              <span
                class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
                data-index="0"
                style="left: 0%;"
              />
              <span
                aria-hidden="true"
                class="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
                data-index="0"
                style="left: 0%;"
              >
                0
              </span>
              <span
                class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
                data-index="1"
                style="left: 100%;"
              />
              <span
                aria-hidden="true"
                class="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
                data-index="1"
                style="left: 100%;"
              >
                3000
              </span>
              <span
                class="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
                data-index="0"
                style="left: 0%;"
              >
                <input
                  aria-label="Always visible"
                  aria-labelledby="input-slider"
                  aria-orientation="horizontal"
                  aria-valuemax="3000"
                  aria-valuemin="0"
                  aria-valuenow="0"
                  data-index="0"
                  max="3000"
                  min="0"
                  step="1"
                  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                  type="range"
                  value="0"
                />
                <span
                  aria-hidden="true"
                  class="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
                >
                  <span
                    class="MuiSlider-valueLabelCircle"
                  >
                    <span
                      class="MuiSlider-valueLabelLabel"
                    >
                      0
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </div>
          <div
            class="MuiBox-root css-30zzyj"
          >
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary css-1tq0l7i-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-invalid="false"
                aria-labelledby="input-slider"
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                max="3000"
                min="0"
                step="1"
                type="number"
                value="0"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-1kj8htw-MuiCardActions-root"
    >
      <button
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButtonBase-root Mui-disabled  css-1knaqv7-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Add to basket
      </button>
    </div>
  </div>
</DocumentFragment>
`;
